### DISCUSSION TITLE
Watch file for changes using vim.uv.new_fs_event()

### DISCUSSION THREAD

**Discussion by Jezda1337** [neovim/neovim#26900] [Q&A]
Hello, I have this code:
```lua
local uv = vim.uv
local event = uv.new_fs_event()
local path = "path to a file"
event:start(path, {
   watch_entry = true,
   stat = true,
  recursive = true,
  }, function(err, filename, events)
  if err then
    print("Error: ", err)
    event:stop()
  end
  print(string.format('Change detected in %s',
	uv.fs_event_getpath(event))) -- get the file name path
  for k, v in pairs(events) do
	print("-------")
	print(k, v)
	print("-------")
  end
  print('file changed:' .. (filename and filename or ''))
end)
```
I have an expectation that the callback function should be triggered every time I make a change to a file. However, it only fires once when I make the first change to the file. I am wondering if this is normal behavior, or if I have misunderstood the documentation. The documentation seems straightforward, so I'm not sure what the issue could be.

output:
```log
"index.html" 12L, 240B written
Change detected in {path to a file}
-------
rename true
-------
file changed:index.html
Change detected in {path to a file}
-------
change true
-------
file changed:index.html
Change detected in {path to a file}
-------
rename true
-------
file changed:index.html
Change detected in {path to a file}
-------
rename true
-------
file changed:index.html
```

**Comment by lewis6991** [comment]
When nvim saves a file it first writes the contents to a temporary file and then renames it.

**Reply by Jezda1337** [reply]
I didn't rename anything, I just `:w` saved a file. Basically when I save the file "rename" is triggered but it also changes get triggered.
 Is this even the right way of watching files for changes? or do you have a better example or idea?

**Reply by lewis6991** [reply]
Yes, when you do `:w`. Nvim writes the buffer to a temporary file and then renames that file to the buffers filename.

**Reply by Jezda1337** [reply]
I understand now why my code was behaving that way. Thank you for clarifying. However, I'm still a little bit confused. Is this a legitimate way of watching files for changes?

**Reply by lewis6991** [reply]
yeah looks fine. Watching files is complicated and behaves differently on different os's.

**Reply by Jezda1337** [reply]
Okay, thanks for the help man
